### PR TITLE
[AIRFLOW-5572] Clear task reschedules when clearing task instances

### PR DESCRIPTION
**Pull request by feluelle**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5572
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Currently when you clear tasks the clear mechanism won't delete reschedule entries related to the task instance from the db.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Review by ashb — APPROVED** [review]

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -94,6 +94,17 @@ def clear_task_instances(tis,
                 # outdated. We make max_tries the maximum value of its
                 # original max_tries or the current task try number.
                 ti.max_tries = max(ti.max_tries, ti.try_number - 1)
+
+            if ti.state == State.UP_FOR_RESCHEDULE:
```
Comment (feluelle) [inline airflow/models/taskinstance.py]: I am not sure about this line. Shouldn't it be clearable unrelated of its state? I put this `if` to only consider task instances that can have reschedules.

Comment (XD-DENG) [reply airflow/models/taskinstance.py]: I believe there may be `TaskReschedule` created, even if CURRENT state of `ti` is not `UP_FOR_RESCHEDULE` (say it **was** `UP_FOR_RESCHEDULE`)?
In that case, this `if` can be removed.

Kindly correct me if I'm wrong.

Comment (milton0825) [reply airflow/models/taskinstance.py]: Yeah I think we can remove the `if`

**Review by milton0825 — APPROVED** [review]

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (feluelle) [reply airflow/models/taskinstance.py]: Do you think I can even move it up right before the `if ti.state == State.RUNNING`?

Or should it happen after the related ti state has been set to `None` ?

**Comment by feluelle** [comment]
Ready for re-Review :) @ashb @milton0825 @XD-DENG 

Waiting for CI to get a green check mark :) - locally it breathed successfully

**Pull request merged**